cmd/k8sagent/unit/agent.go

```diff
@@ -0,0 +1,270 @@
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/featureflag"
+	"github.com/juju/gnuflag"
+	"github.com/juju/loggo"
+	"github.com/juju/names/v4"
+	"github.com/juju/utils/voyeur"
+	"github.com/juju/worker/v2"
+	"github.com/juju/worker/v2/dependency"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/juju/juju/agent"
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/api/uniter"
+	jujucmd "github.com/juju/juju/cmd"
+	jujudagent "github.com/juju/juju/cmd/jujud/agent"
+	cmdutil "github.com/juju/juju/cmd/jujud/util"
+	"github.com/juju/juju/core/machinelock"
+	"github.com/juju/juju/upgrades"
+	jujuversion "github.com/juju/juju/version"
+	jworker "github.com/juju/juju/worker"
+	"github.com/juju/juju/worker/gate"
+	"github.com/juju/juju/worker/introspection"
+	"github.com/juju/juju/worker/logsender"
+	"github.com/juju/juju/worker/upgradesteps"
+)
+
+var logger = loggo.GetLogger("juju.cmd.k8sagent.unit")
+
+type k8sUnitAgent struct {
+	cmd.CommandBase
+	jujudagent.AgentConf
+	configChangedVal *voyeur.Value
+	// Unit command of k8sagent only knows application name but not unit name.
+	// It will configure out the unit name from agent.conf file by itself.
+	applicationName string
+	clk             clock.Clock
+	runner          *worker.Runner
+	bufferedLogger  *logsender.BufferedLogWriter
+	setupLogging    func(agent.Config) error
+	ctx             *cmd.Context
+	dead            chan struct{}
+	errReason       error
+	machineLock     machinelock.Lock
+
+	// Used to signal that the upgrade worker will not
+	// reboot the agent on startup because there are no
+	// longer any immediately pending agent upgrades.
+	initialUpgradeCheckComplete gate.Lock
+	preUpgradeSteps             upgrades.PreUpgradeStepsFunc
+	upgradeComplete             gate.Lock
+
+	prometheusRegistry *prometheus.Registry
+}
+
+func New(ctx *cmd.Context, bufferedLogger *logsender.BufferedLogWriter) (cmd.Command, error) {
+	prometheusRegistry, err := jujudagent.NewPrometheusRegistry()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &k8sUnitAgent{
+		AgentConf:                   jujudagent.NewAgentConf(""),
+		configChangedVal:            voyeur.NewValue(true),
+		ctx:                         ctx,
+		clk:                         clock.WallClock,
+		dead:                        make(chan struct{}),
+		initialUpgradeCheckComplete: gate.NewLock(),
+		bufferedLogger:              bufferedLogger,
+		prometheusRegistry:          prometheusRegistry,
+		preUpgradeSteps:             upgrades.PreUpgradeSteps,
+	}, nil
+}
+
+// Info returns a description of the command.
+func (c *k8sUnitAgent) Info() *cmd.Info {
+	return jujucmd.Info(&cmd.Info{
+		Name:    "unit",
+		Purpose: "starting a k8s agent",
+	})
+}
+
+// SetFlags implements Command.
+func (c *k8sUnitAgent) SetFlags(f *gnuflag.FlagSet) {
+	c.AgentConf.AddFlags(f)
+	f.StringVar(&c.applicationName, "application-name", "", "name of the application")
```

Comment:
I think by the time we are starting the unit, it has an identity, so we want the unit name, not the application name.

We talked about having the init command the thing that will write down the agent.conf file for the k8s agent.

I also think we should take a --data-dir and --log-dir arg because the default linux directories won't be right for k8s.  I think that is better than hard coding values in here as args > hard coding.

Reply:
You are right, the init command prepares the agent.conf with tag == unitTag inside.
Here the unit command reads the agent.conf to get its own unitTag.

We either put the agent.conf to somewhere the path does have application name information at all, for example, `/juju/`.
But current PR chooses to put at `/var/lib/juju/agents/<application-tag>/agent.conf`.

The `c.AgentConf.AddFlags(f)` adds `--data-dir` option already.
I found `jujud machine|unit|caasoperator`does not have `--log-dir` option, do you want we to add this new option for k8sagent only or?

Reply:
I think we may need to add them later. I'd only add them for the k8s uniter, but let's wait and see as we don't know exactly how things are going to end up sitting in the end. 

If we do end up running this in a sidecar, then perhaps it may live in /var/lib/juju, but if we run inside the workload container, it may be /juju/agent or something. Since we don't know yet, let's not add it in yet.